### PR TITLE
Implement subfield acl check

### PR DESCRIPTION
**Pull request by pranjalssh**
Added an option to enable fine grained column subfield acl check. If we query col.x, we send "col.x" to permission server for access control checks, rather than just "col".

Also noticed that UtilizedColumnReferenceAnalyzer had a bug with dereference expressions. I fixed it(one liner) and I guess people weren't relying on it.

After fixing it, UtilizedColumnReferenceAnalyzer implementation still doesn't work perfectly with subfields, as it doesn't prune unused subfields(added a test with explanation), but the implementation should be more than enough to push it like this, especially given the fact that it was already buggy before.

```
== RELEASE NOTES ==

General Changes

* Allow connectors to support column subfield access control checks. Connectors can specify subfield access control for row-type columns through ``checkCanSelectFromColumns()`` in SPI.

```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 1]
```diff
@@ -760,18 +765,23 @@ public JoinUsingAnalysis getJoinUsing(Join node)
...
 
-    public void addTableColumnReferences(AccessControl accessControl, Identity identity, Multimap<QualifiedObjectName, String> tableColumnMap)
+    public void addTableColumnAndSubfieldReferences(AccessControl accessControl, Identity identity, Multimap<QualifiedObjectName, Subfield> tableColumnMap)
     {
         AccessControlInfo accessControlInfo = new AccessControlInfo(accessControl, identity);
-        Map<QualifiedObjectName, Set<String>> references = tableColumnReferences.computeIfAbsent(accessControlInfo, k -> new LinkedHashMap<>());
+        Map<QualifiedObjectName, Set<String>> columnReferences = tableColumnReferences.computeIfAbsent(accessControlInfo, k -> new LinkedHashMap<>());
         tableColumnMap.asMap()
-                .forEach((key, value) -> references.computeIfAbsent(key, k -> new HashSet<>()).addAll(value));
+                .forEach((key, value) -> columnReferences.computeIfAbsent(key, k -> new HashSet<>()).addAll(value.stream().map(subfield -> subfield.getRootName()).collect(toImmutableSet())));
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java:773]: nit:
s/`subfield -> subfield.getRootName()`/`Subfield::getRootName`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 2]
```diff
@@ -789,29 +799,60 @@ public void addUtilizedTableColumnReferences(AccessControlInfo accessControlInfo
...
+        Map<AccessControlInfo, Map<QualifiedObjectName, Set<Subfield>>> references;
+        if (!isCheckAccessControlWithSubfields(session)) {
+            references = (isCheckAccessControlOnUtilizedColumnsOnly(session) ? utilizedTableColumnReferences : tableColumnReferences)
+                    .entrySet()
+                    .stream()
+                    .collect(toImmutableMap(
+                            Map.Entry::getKey,
+                            accessControlEntry -> accessControlEntry.getValue().entrySet().stream().collect(toImmutableMap(
+                                    Map.Entry::getKey,
+                                    tableEntry -> tableEntry.getValue().stream().map(name -> new Subfield(name)).collect(toImmutableSet())))));
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java:813]: nit
s/`name -> new Subfield(name)`/`Subfield::new`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 3]
```diff
@@ -789,29 +799,60 @@ public void addUtilizedTableColumnReferences(AccessControlInfo accessControlInfo
...
 
-    public Map<AccessControlInfo, Map<QualifiedObjectName, Set<String>>> getTableColumnReferencesForAccessControl(Session session)
+    public Map<AccessControlInfo, Map<QualifiedObjectName, Set<Subfield>>> getTableColumnAndSubfieldReferencesForAccessControl(Session session)
     {
-        Map<AccessControlInfo, Map<QualifiedObjectName, Set<String>>> references = isCheckAccessControlOnUtilizedColumnsOnly(session) ? utilizedTableColumnReferences : tableColumnReferences;
+        Map<AccessControlInfo, Map<QualifiedObjectName, Set<Subfield>>> references;
+        if (!isCheckAccessControlWithSubfields(session)) {
+            references = (isCheckAccessControlOnUtilizedColumnsOnly(session) ? utilizedTableColumnReferences : tableColumnReferences)
+                    .entrySet()
+                    .stream()
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java:808]: nit: put `.entrySet().stream()` at the end of the last line. We usually break a new line at `stream()`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 4]
```diff
@@ -440,11 +445,11 @@ private Type handleResolvedField(Expression node, FieldId fieldId, Field field,
         protected Type visitDereferenceExpression(DereferenceExpression node, StackableAstVisitorContext<Context> context)
         {
             QualifiedName qualifiedName = DereferenceExpression.getQualifiedName(node);
+            Scope scope = context.getContext().getScope();
 
             // Handle qualified name
             if (qualifiedName != null) {
                 // first, try to match it to a column name
-                Scope scope = context.getContext().getScope();
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java:447]: accidental change?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 5]
```diff
@@ -1336,6 +1345,77 @@ public Type visitGroupingOperation(GroupingOperation node, StackableAstVisitorCo
...
+
+        private boolean isTopMostReference(Expression node, StackableAstVisitorContext<Context> context)
+        {
+            if (!context.getPreviousNode().isPresent()) {
+                return true;
+            }
+            return !isDereferenceOrSubscript((Expression) context.getPreviousNode().get());
+        }
+
+        private void addColumnSubFieldReferences(Expression node, StackableAstVisitorContext<Context> context)
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java:1361]: nit lower case f: Subfield

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 6]
```diff
@@ -1336,6 +1345,77 @@ public Type visitGroupingOperation(GroupingOperation node, StackableAstVisitorCo
...
+            Expression childNode = node;
+            List<PathElement> columnDereferences = new ArrayList<>();
+            while (true) {
+                if (childNode instanceof SubscriptExpression) {
+                    SubscriptExpression subscriptExpression = (SubscriptExpression) childNode;
+                    childNode = subscriptExpression.getBase();
+                    Type baseType = expressionTypes.get(NodeRef.of(childNode));
+                    if (baseType == null || !(baseType instanceof RowType)) {
+                        continue;
+                    }
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java:1377]: `!(baseType instanceof RowType)` implies `baseType == null`

**Review by highker — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java (hunk 6 above)
Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java:1377]: Also remember we have indexed access to fields x.a is same as x[n] if a is the n-th field in the struct.

**Comment by kaikalur** [comment]
Also remember we have indexed access to fields x[n]is same as x.a if a is the n-th field in the struct.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java (hunk 6 above)
Comment (pranjalssh) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java:1377]: Why? This is for subscript, it could be array/map

**Pull request merged**